RFC-0026-logging-system.md

```diff
@@ -0,0 +1,486 @@
+    [PyTorch feature classifications](https://pytorch.org/blog/pytorch-feature-classification-changes/).
+  - TODO: This warning type might not be very useful--find out if we really
+    want this
+
+* **`c10::NondeterministicWarning`** - Python `torch.NondeterministicWarning`
+  - Emitted when `torch.use_deterministic_algorithms(True)` and
+    `torch.set_deterministic_debug_mode('warn')` are set, and a
+    nondeterministic operation is called.
+
+* **`c10::DeprecationWarning`** - Python `DeprecationWarning`
+  - Emitted when a deprecated function is called.
+  - TODO: `DeprecationWarning`s are ignored by default in Python, so we may
+    actually want to use a different Python class for this.
+
+
+#### Info message classes:
+  
+* **`c10::Info`** - Python `torch.Info`
+  - Default info class. Other info classes inherit from it.
+
+
+### Message APIs
+
+In order to emit messages, developers can use the APIs defined in this section.
+
+These APIs all have a variable length argument list, `...` in C++ and `*args`
+in Python. When a message is emitted, these arguments are concatenated into
+a string, and the string becomes the body of the message.
+
+In C++, the arguments in `...` must all have the `std::ostream& operator<<`
+function defined so that they can be concatenated.
+
+In Python, each element in `*args` must either have a `__str__` function or it
+must be a callable that, when called, produces another object that has
+a `__str__` fuction. Providing the body of a message as a callable can provide
+better performance in cases where the message would not be emitted, as in
+`torch.check(True, lambda: expensive_function())` if `cond == True`, since the
+`expensive_function()` would not be called in that case.
+
+
+#### Error APIs
+
+The APIs for raising errors all check a boolean condition, the `cond` argument
+in the following signatures, and throw an error if that condition is false.
+
+The error APIs are listed below, with the C++ signature on the left and the
+corresponding Python signature on the right.
+
+**`TORCH_CHECK(cond, ...)`** - `torch.check(cond, *args)`
+  - C++ error: `c10::Error`
+  - Python error: `RuntimeError`
+
+**`TORCH_CHECK_INDEX(cond, ...)`** - `torch.check_index(cond, *args)`
+  - C++ error: `c10::IndexError`
+  - Python error: `IndexError`
+
+**`TORCH_CHECK_VALUE(cond, ...)`** - `torch.check_value(cond, *args)`
+  - C++ error: `c10::ValueError`
+  - Python error: `IndexError`
+
+**`TORCH_CHECK_TYPE(cond, ...)`** - `torch.check_type(cond, *args)`
+  - C++ error: `c10::TypeError`
+  - Python error: `TypeError`
+
+**`TORCH_CHECK_NOT_IMPLEMENTED(cond, ...)`** - `torch.check_not_implemented(cond, *args)`
+  - C++ error: `c10::NotImplementedError`
+  - Python error: `NotImplementedError`
+
+**`TORCH_CHECK_WITH(error_t, cond, ...)`** - `torch.check_with(error_type, cond, *args)`
+  - C++ error: Specified by `error_t` argument
+  - Python error: Specified by `error_type` argument
```

Comment:
We want to change how asserts are written to support vmap (and/or Tensor Subclasses) over functions that call asserts (e.g. https://github.com/pytorch/pytorch/pull/81767, cc @kshitij12345). In order to support those, the assert should go through the PyTorch dispatcher. Since we are re-thinking how asserts in PyTorch work, this is a good opportunity to figure out how we can have asserts as PyTorch operators.

## Some background
folks write functions in C++ like the following:
```
Tensor safe_log(const Tensor& x) {
  TORCH_CHECK((x > 0).all(), "requires x > 0");
  return x.log()
}
```
If someone is doing `vmap(safe_log)(torch.tensor([-1, 1, 2))`, then `(x > 0).all()` is a Tensor that has multiple values. A Tensor that has multiple values cannot be coerced to a single boolean value, and the above errors out.

## A sketch of a solution

For synchronous asserts, we would want two extension points. One to actually check the condition, and the other to raise an error message. So `TORCH_CHECK(cond, ...)` would be implemented like:
```
if (at::is_true(cond)) {
  msg = ...
  at::fail(msg)
}
```

Reply:
@zou3519, I started looking into adding vmap support to `TORCH_CHECK`, and I realized that it doesn't actually support tensor inputs at all, not even non-batched tensors. For example, if I do this:

```cpp
TORCH_CHECK((at::zeros(10).eq(0)).all());
```

I get this build error:

```
/work2/kurtamohler/development/pytorch-1/c10/util/Exception.h:506:29: error: no match for 'operator!' (operand type is 'at::Tensor')
  506 |   if (C10_UNLIKELY_OR_CONST(!(cond))) {            \
```

Without `at::Tensor::operator!`, we currently have to do something like this instead:

```cpp
TORCH_CHECK((at::zeros(10).eq(0)).all().item<bool>());
```

So I like the idea of adding support for tensor inputs to `TORCH_CHECK`.

One way, could be to use an `at::is_true` function, like you mentioned. However, I'm not sure exactly how to do that. Since `TORCH_CHECK` is defined in `c10`, it doesn't have access to the `at` namespace. But of course, since it's a macro, it's just doing substitution, so maybe we could have an `#ifdef` that checks whether `at::is_true` is available before using it, like this:

```cpp
#ifdef AT_IS_TRUE_DEFINED
#define CHECK_COND(cond) at::is_true(cond)
#else
#define CHECK_COND(cond) (cond)
#endif
```

The `#define AT_IS_TRUE_DEFINED` would be next to wherever the declaration of `at::is_true` is, and then we can use `CHECK_COND` in the definition of `TORCH_CHECK`.

But when I tried this, I couldn't figure out which file to define `at::is_true` such that its available for any `TORCH_CHECK` call that might use tensors. It also just seems like kind of a sketchy solution.

Another option is to just define `at::Tensor::operator!`. But in order to make `TORCH_CHECK` work correctly when `cond` is a batched tensor, the actual meaning of `at::Tensor::operator!` would have to be like `!physical_tensor.any()`. That seems like an inconsistency to me, but maybe it's actually fine, I'm not sure.

Maybe there's another better way to do this.

Reply:
Is there any reason you have to use TORCH_CHECK for this

Reply:
As I understood it, the suggestion was to make `TORCH_CHECK((x > 0).all(), "requires x > 0");`, for example, work for both batched and non-batched tensors

Reply:
I guess, for something like this where we're syncing, I think it is fine if we force the call sites to change what they look like. There's not many of this kind of check.

Reply:
I'm fine with having something like a TORCH_CHECK_TENSOR that accepts a Tensor to sidestep the issues above. We control what gets written in our C++ internals so we can always refactor if we're not satisfied.

In Python, is it correct that we want `torch.check` to be public API? If so it needs more scrutiny. My initial thought is that torch.check should do both TORCH_CHECK and TORCH_CHECK_TENSOR (there aren't technical reasons in Python for us to split those up)

Reply:
I don't think the Python API needs to be public. I can make it private to start with, and we can make it public if we ever find a reason to